cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java

```diff
@@ -151,29 +154,37 @@ public void createNamespace(NamespaceMeta metadata) throws AlreadyExistsExceptio
    * Deletes the specified namespace
    *
    * @param namespaceId the {@link Id.Namespace} of the specified namespace
+   * @throws NamespaceCannotBeDeletedException if the specified namespace cannot be deleted
    * @throws NotFoundException if the specified namespace does not exist
    */
-  public void deleteNamespace(Id.Namespace namespaceId) throws NotFoundException, IOException {
-    //TODO: CDAP-870, CDAP-1427. Delete should be in a single transaction.
+  public void deleteNamespace(Id.Namespace namespaceId) throws NamespaceCannotBeDeletedException, NotFoundException {
+    // TODO: CDAP-870, CDAP-1427: Delete should be in a single transaction.
+    if (store.getNamespace(namespaceId) == null) {
+      throw new NotFoundException(NAMESPACE_ELEMENT_TYPE, namespaceId.getId());
+    }
+    // TODO: CDAP-870: Check if apps are running and abort if they are
     // Delete Preferences associated with this namespace
     preferencesStore.deleteProperties(namespaceId.getId());
 
     // Delete all dashboards associated with this namespace
     dashboardStore.delete(namespaceId.getId());
 
+    // move data directory and try to remove HBase namespace
+    try {
+      dsFramework.deleteAllInstances(namespaceId);
+      dsFramework.deleteAllModules(namespaceId);
+      dsFramework.deleteNamespace(namespaceId);
+    } catch (DatasetManagementException e) {
+      throw new NamespaceCannotBeDeletedException(namespaceId.getId(), e);
+    } catch (IOException e) {
+      throw new NamespaceCannotBeDeletedException(namespaceId.getId(), e);
+    }
+
+    // TODO: CDAP-870: Delete streams, queues, apps
     // Store#deleteNamespace already checks for existence
     NamespaceMeta deletedNamespace = store.deleteNamespace(namespaceId);
     if (deletedNamespace == null) {
```

Comment:
Same. Don't feel too strongly about it though, just an extra level of checking. I agree though that there could be other side effects of it not being atomic, that this check won't prevent.